pymc_marketing/customer_choice/nested_logit.py

```diff
@@ -0,0 +1,783 @@
+                w_nest = pm.math.zeros((len(coords["obs"]), len(coords["alts"])))
+                betas_fixed_ = None
+            else:
+                W_data = pm.Data("W", W, dims=("obs", "fixed_covariates"))
+                betas_fixed_ = pm.Normal(
+                    "betas_fixed_", 0, 1, dims=("alts", "fixed_covariates")
+                )
+                betas_fixed = pm.Deterministic(
+                    "betas_fixed",
+                    pt.set_subtensor(betas_fixed_[-1, :], 0),
+                    dims=("alts", "fixed_covariates"),
+                )
+                w_nest = pm.math.dot(W_data, betas_fixed.T)
+            X_data = pm.Data("X", X, dims=("obs", "alts", "alt_covariates"))
+            y_data = pm.Data("y", y, dims="obs")
+
+            # Compute utility as a dot product
+            alphas = pt.set_subtensor(alphas[-1], 0)
+            u = alphas + pm.math.dot(X_data, betas)
+            U = pm.Deterministic("U", w_nest + u, dims=("obs", "alts"))
+
+            ## Mid Level
+            if "mid" in nest_indices.keys():
+                cond_prob_m, nest_prob_m = self.make_P_nest(
+                    U, W, betas_fixed_, lambdas_nests, "mid"
+                )
+
+                ## Construct Paths Bottom -> Up
+                child_nests = {}
+                path_prods_m: dict[str, list] = {}
+                ordered = [
+                    (key, min(vals)) for key, vals in nest_indices["mid"].items()
+                ]
+                middle_nests = [x[0] for x in sorted(ordered, key=lambda x: x[1])]
+                for idx, n in enumerate(middle_nests):
+                    is_last = idx == len(middle_nests) - 1
+                    parent, child = n.split("_")
+                    P_nest = nest_prob_m[n]
+                    P_y_given_nest = cond_prob_m[n]["P_y_given"]
+                    prod = pm.Deterministic(
+                        f"prod_{n}_m", (P_nest[:, pt.newaxis] * P_y_given_nest)
+                    )
+                    if parent in path_prods_m:
+                        path_prods_m[parent].append(prod)
+                    else:
+                        path_prods_m[parent] = []
+                        path_prods_m[parent].append(prod)
+                    if is_last:
+                        P_ = pm.Deterministic(
+                            f"P_{parent}_children",
+                            pm.math.concatenate(path_prods_m[parent], axis=1),
+                        )
+                        child_nests[parent] = P_
+            else:
+                child_nests = {}
+
+            ## Top Level
+            cond_prob_t, nest_prob_t = self.make_P_nest(
+                U, W, betas_fixed_, lambdas_nests, "top"
+            )
+
+            path_prods_t = []
+            ordered = [(key, min(vals)) for key, vals in nest_indices["top"].items()]
+            top_nests = [x[0] for x in sorted(ordered, key=lambda x: x[1])]
+            for _idx, n in enumerate(top_nests):
```

Comment:
An empty line before this line to make the code more readable?

Reply:
done